### PR TITLE
Receive post requests

### LINKED ISSUE #10: Add receiving a list of emails and returning a successful status code
The application should accept a POST request with the following format

```
{
	'users':[
		{'name':'jenny jones', 'email':'jenny@jones.com'}, 
		{'name':'amelia amerson', 'email':'amelia@amerson.com'}, 
		{'name':'jane smith', 'email':'jane@smith.com'}
	],
	'master':{'name':'jenny jones','email':'jenny@jones.com'} 
}
```

### PR DESCRIPTION
**Pull request by joshuarmil**
Closes #10

### CODE REVIEW

**Comment by joshuarmil** [comment]
@ptravers How does this look? Is it still simple enough that we want this handled in main.js? Also, should I be making the request in this same commit? I didn't see any issue for making a user form and wasn't sure how to validate the post request success, so I found a sample form and used that.

Also, I see that there is a conflict with package.json, but I'm not sure how to resolve, I added the body-parser dependency (not actually needed i think, because it's in express by default from my understanding)

**Comment by joshuarmil** [comment]
Forget if we covered this, should I make a separate branch for each issue, or can I combine if they are dependent on each other?

File: package.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "start": "node back-end/bin/www"
+  },
+  "keywords": [
+    "Secret",
+    "Santa",
+    "Polyanna"
+  ],
+  "author": "Josh Miller and Peter Travers",
```
Comment (ptravers) [inline package.json]: This is all you!

File: back-end/main.js [hunk 2]
```diff
@@ -16,4 +35,13 @@ app.get('/status', function(req, res) {
 	res.end();
 });
 
+app.post('/post', function(req, res) {
+	console.log(req.body);
+	transporter.sendMail(mailOptions, function(err, info) {
```
Comment (ptravers) [inline back-end/main.js]: So rather than try to send the email we want to just check if the body is the correct structure

File: back-end/main.js [hunk 3]
```diff
@@ -1,9 +1,28 @@
...
+		pass: process.env.EMAILPASS
+	}
+});
+
+const mailOptions = {
+	from: process.env.EMAILADDRESS,
+	to: process.env.RECIPIENT,
+	subject: 'Polly Anna Time!',
+	html: '<p>Hi, how are you?</p>'
+};
```
Comment (ptravers) [inline back-end/main.js]: I think you can save this for the later ticket

**Comment by ptravers** [comment]
@joshuarmil generally one per issue

File: back-end/main.js [hunk 4]
```diff
@@ -16,4 +35,13 @@ app.get('/status', function(req, res) {
 	res.end();
 });
 
+app.post('/post', function(req, res) {
```
Comment (low-ghost) [inline back-end/main.js]: not terribly important, but POST is the html method that will be accepted. It's convention to name your route as a resource and then have various actions against it. In this case what are we creating via calling POST? just the email? If that were the case it'd probably be at '/email', but it'll probably be things more like '/invite' or '/gift' that would send emails as part of their logic. Those would also have other meaningful html 'verbs' like 'get a list of all gifts via GET' or 'delete that one gift via DELETE /gift/2'

You're just building out/testing functionality here so call it whatever you like, I'm just typing for the hell of it

**Comment by low-ghost** [comment]
How issues/tickets relate to git is really getting into the weeds of software development and its never a truly solved issue. Generally, you want to have a branch for every issue. The ideal is to plan out features so that they aren't dependent on each other, or at least to get them reviewed and merged before you have a chain of PR's. That's one way to deal with it though, make a PR that targets the branch that is already in PR, and a PR that targets that one etc. Another is 'feature branches' which are like a 'develop' branch but are about building out larger feature sets that will inevitably have dependent code. Then you can merge more freely to that branch and review when that branch goes into develop. It's a whole thing though, probably not worth thinking too much about.

**Comment by low-ghost** [comment]
Looks good to me though. Important thing is to try
```sh
curl -X POST -d '{"some": "data"}'  http://localhost:3000/post
```
or the javascript equivalent, depending on your libs that'd be
```js
fetch('http://localhost:3000/post', { method: 'POST', body: JSON.stringify({ some: 'data' }) })
```
and make sure that you're getting the content in your backend terminal due to your `console.log`, and maybe even sending an email 😃

File: back-end/main.js (hunk 2 above)
Comment (joshuarmil) [reply back-end/main.js]: Is there a standard way to validate structure or validate requests in general? I found stuff like this:https://medium.com/@xoor/building-a-node-js-rest-api-7-request-validation-43c144300d06 but wasn't sure if that was overkill?

File: back-end/main.js (hunk 4 above)
Comment (joshuarmil) [reply back-end/main.js]: ah okay that makes sense, i will adjust. should i be using arrow functions everywhere here as previously discussed?

**Comment by joshuarmil** [comment]
re: fetch/curl, i spent way too long trying to figure out the best way to send requests, and i knew there had to be something simple, but everything i found was way overboard, using jquery and stuff, so i landed on the xmlhttprequest thing, but is that still overkill? i should be using fetch for pretty much everything at this point?

**Comment by joshuarmil** [comment]
> How issues/tickets relate to git is really getting into the weeds of software development and its never a truly solved issue. Generally, you want to have a branch for every issue. The ideal is to plan out features so that they aren't dependent on each other, or at least to get them reviewed and merged before you have a chain of PR's. That's one way to deal with it though, make a PR that targets the branch that is already in PR, and a PR that targets that one etc. Another is 'feature branches' which are like a 'develop' branch but are about building out larger feature sets that will inevitably have dependent code. Then you can merge more freely to that branch and review when that branch goes into develop. It's a whole thing though, probably not worth thinking too much about.

So maybe rather than make PRs for each of these issues, I should just build up the branch as it goes and then wait til it truly needs to be merged to make the PR?

**Comment by ptravers** [comment]
@joshuarmil Really these tickets should be subtickets of a ticket. My goal when writing them was to provide a set of clean steps that should follow in order. I think each should be mergable in it's own right since there should be no user impact. @low-ghost might disagree but in my experience a backend endpoint will ship having been appropriately tested. Then a front end connection to it will ship.

**Comment by joshuarmil** [comment]
@ptravers the steps have been good and have definitely helped me get a direction to follow so i don't get hung up not knowing what the next best focus is. 

Should i turn off the endpoint at this point, or keep it open for testing? seems like that should be the next step any way. Makes sense about doing the back-end first, i only jumped to front end because i wasn't sure how to best test the end point, not realizing about fetch.

For the feature branching, maybe I'm just not seeing this clearly, being new, but feels like aside from front end and back end, there is really only one feature, aside from something like the login. The rest, as you've outlined, seems like a logical progression to a working site

**Comment by ptravers** [comment]
@joshuarmil right so you would consider this to be a feature. I wouldn't worry about that that only matters for production applications and it's a purely stylistic thing. I should have left it at what bagwell said. what you're doing is great. this is all in the stylistic weeds as bagwell mentioned.

File: back-end/main.js (hunk 2 above)
Comment (joshuarmil) [reply back-end/main.js]: @ptravers I was also looking at JSON Schema and express has some middleware express-validator which can check each field, but it looked like it added a lot to the complexity

Comment (ptravers) [reply back-end/main.js]: I guess you don't really need to check JSON if you're writing JS. @low-ghost does that seem right to you?

**Pull request merged**